--- a/security/selinux/selinuxfs.c	2026-08-09 20:25:30.000000000 +0200
+++ b/security/selinux/selinuxfs.c	2026-08-12 08:03:48.138218306 +0200
@@ -654,13 +654,46 @@
 static ssize_t sel_write_checkreqprot(struct file *file, const char __user *buf,
 				      size_t count, loff_t *ppos)
 {
-	/*
-	 * Setting checkreqprot is no longer supported, see
-	 * https://github.com/SELinuxProject/selinux-kernel/wiki/DEPRECATE-checkreqprot
-	 */
-	pr_err_once("SELinux: %s (%d) wrote to checkreqprot. This is no longer supported.\n",
-		    current->comm, current->pid);
-	return count;
+	char *page;
+	ssize_t length;
+	unsigned int new_value;
+
+	length = avc_has_perm(current_sid(), SECINITSID_SECURITY,
+			      SECCLASS_SECURITY, SECURITY__SETCHECKREQPROT,
+			      NULL);
+	if (length)
+		return length;
+
+	if (count >= PAGE_SIZE)
+		return -ENOMEM;
+
+	/* No partial writes. */
+	if (*ppos != 0)
+		return -EINVAL;
+
+	page = memdup_user_nul(buf, count);
+	if (IS_ERR(page))
+		return PTR_ERR(page);
+
+	if (sscanf(page, "%u", &new_value) != 1) {
+		length = -EINVAL;
+		goto out;
+	}
+	length = count;
+
+	if (new_value) {
+		char comm[sizeof(current->comm)];
+
+		memcpy(comm, current->comm, sizeof(comm));
+		pr_err("SELinux: %s (%d) set checkreqprot to 1. This is no longer supported.\n",
+		       comm, current->pid);
+	}
+
+	selinux_ima_measure_state();
+
+out:
+	kfree(page);
+	return length;
 }
 static const struct file_operations sel_checkreqprot_ops = {
 	.read		= sel_read_checkreqprot,
