--- a/drivers/tty/vt/vt.c	2020-04-20 19:28:38.000000000 +0200
+++ b/drivers/tty/vt/vt.c	2020-05-01 08:03:36.059292624 +0200
@@ -3411,7 +3411,7 @@
 		panic("Couldn't register /dev/tty0 driver\n");
 	tty0dev = device_create_with_groups(tty_class, NULL,
 					    MKDEV(TTY_MAJOR, 0), NULL,
-					    vt_dev_groups, "tty0");
+					    vt_dev_groups, "brunch-tty0");
 	if (IS_ERR(tty0dev))
 		tty0dev = NULL;
 
