diff -ruN a/drivers/pci/pcie/aer.c b/drivers/pci/pcie/aer.c
--- a/drivers/pci/pcie/aer.c	2019-07-21 09:03:18.000000000 +0200
+++ b/drivers/pci/pcie/aer.c	2019-10-19 18:54:56.557377247 +0200
@@ -24,6 +24,7 @@
 #include <linux/delay.h>
 #include <linux/kfifo.h>
 #include <linux/slab.h>
+#include <linux/surface_devices_dmi.h>
 #include <acpi/apei.h>
 #include <ras/ras_event.h>
 
@@ -35,6 +36,8 @@
 #define AER_MAX_TYPEOF_COR_ERRS		16	/* as per PCI_ERR_COR_STATUS */
 #define AER_MAX_TYPEOF_UNCOR_ERRS	26	/* as per PCI_ERR_UNCOR_STATUS*/
 
+static const struct dmi_system_id devices[] = surface_go_devices;
+
 struct aer_err_source {
 	unsigned int status;
 	unsigned int id;
@@ -120,7 +123,7 @@
 
 bool pci_aer_available(void)
 {
-	return !pcie_aer_disable && pci_msi_enabled();
+	return !pcie_aer_disable && pci_msi_enabled() && !dmi_check_system(devices);
 }
 
 #ifdef CONFIG_PCIE_ECRC
