--- a/drivers/net/wireless/ath/ath10k/mac.c	2020-12-16 07:36:16.000000000 +0100
+++ b/drivers/net/wireless/ath/ath10k/mac.c	2020-12-28 23:14:04.401861885 +0100
@@ -8968,9 +8968,9 @@
 	} else if (rate == mcs_rate[index].supported_rate[3]) {
 		*bw |= RATE_INFO_BW_40;
 		*flags |= RATE_INFO_FLAGS_SHORT_GI;
-	} else {
-		ath10k_warn(ar, "invalid ht params rate %d 100kbps nss %d mcs %d",
-			    rate, nss, mcs);
+	//} else {
+	//	ath10k_warn(ar, "invalid ht params rate %d 100kbps nss %d mcs %d",
+	//		    rate, nss, mcs);
 	}
 }
 
