diff -uN a/drivers/hid/hid-multitouch.c b/drivers/hid/hid-multitouch.c
--- a/drivers/hid/hid-multitouch.c	2019-11-18 05:36:32.000000000 +0100
+++ b/drivers/hid/hid-multitouch.c	2020-01-09 07:12:55.124145556 +0100
@@ -191,6 +191,7 @@
 #define MT_CLS_WIN_8				0x0012
 #define MT_CLS_EXPORT_ALL_INPUTS		0x0013
 #define MT_CLS_WIN_8_DUAL			0x0014
+#define MT_CLS_WIN_8_FORCE_GET_FEATURE		0x0015
 
 /* vendor specific classes */
 #define MT_CLS_3M				0x0101
@@ -279,6 +280,14 @@
 			MT_QUIRK_CONTACT_CNT_ACCURATE |
 			MT_QUIRK_WIN8_PTP_BUTTONS,
 		.export_all_inputs = true },
+	{ .name = MT_CLS_WIN_8_FORCE_GET_FEATURE,
+		.quirks = MT_QUIRK_ALWAYS_VALID |
+			MT_QUIRK_IGNORE_DUPLICATES |
+			MT_QUIRK_HOVERING |
+			MT_QUIRK_CONTACT_CNT_ACCURATE |
+			MT_QUIRK_STICKY_FINGERS |
+			MT_QUIRK_WIN8_PTP_BUTTONS |
+			MT_QUIRK_FORCE_GET_FEATURE },
 
 	/*
 	 * vendor specific classes
@@ -1973,6 +1982,14 @@
 		HID_USB_DEVICE(USB_VENDOR_ID_LG,
 			USB_DEVICE_ID_LG_MELFAS_MT) },
 
+	/* Microsoft Surface Go type cover */
+	{ .driver_data = MT_CLS_WIN_8_FORCE_GET_FEATURE,
+		HID_DEVICE(BUS_USB, HID_GROUP_MULTITOUCH_WIN_8, 0x045e, 0x096f) },
+
+	/* Microsoft Surface Go 2 type cover */
+	{ .driver_data = MT_CLS_WIN_8_FORCE_GET_FEATURE,
+		HID_DEVICE(BUS_USB, HID_GROUP_MULTITOUCH_WIN_8, 0x045e, 0x09b5) },
+
 	/* MosArt panels */
 	{ .driver_data = MT_CLS_CONFIDENCE_MINUS_ONE,
 		MT_USB_DEVICE(USB_VENDOR_ID_ASUS,
