diff -ruN a/block/genhd.c b/block/genhd.c
--- a/block/genhd.c	2019-06-23 09:02:58.000000000 +0200
+++ b/block/genhd.c	2019-06-23 15:15:31.963451019 +0200
@@ -37,6 +37,8 @@
 #include "blk.h"
 #include "blk-rq-qos.h"
 
+static int fake_removable = -1;
+
 static struct kobject *block_depr;
 
 /*
@@ -1080,10 +1081,28 @@
 {
 	struct gendisk *disk = dev_to_disk(dev);
 
+	if (fake_removable != -1)
+		return sprintf(buf, "%d\n", fake_removable);
+
 	return sprintf(buf, "%d\n",
 		       (disk->flags & GENHD_FL_REMOVABLE ? 1 : 0));
 }
 
+static ssize_t disk_removable_store(struct device *dev,
+		struct device_attribute *attr, const char *buf, size_t count)
+{
+	int value;
+
+	sscanf(buf, "%d", &value);
+
+	if (value == 0 || value == 1) {
+		fake_removable = value;
+		return count;
+	}
+
+	return -EINVAL;
+}
+
 static ssize_t disk_hidden_show(struct device *dev,
 				   struct device_attribute *attr, char *buf)
 {
@@ -1129,7 +1148,7 @@
 
 static DEVICE_ATTR(range, 0444, disk_range_show, NULL);
 static DEVICE_ATTR(ext_range, 0444, disk_ext_range_show, NULL);
-static DEVICE_ATTR(removable, 0444, disk_removable_show, NULL);
+static DEVICE_ATTR(removable, 0644, disk_removable_show, disk_removable_store);
 static DEVICE_ATTR(hidden, 0444, disk_hidden_show, NULL);
 static DEVICE_ATTR(ro, 0444, disk_ro_show, NULL);
 static DEVICE_ATTR(size, 0444, part_size_show, NULL);
