--- a/arch/x86/kernel/cpu/bugs.c	2020-08-29 06:32:44.000000000 +0200
+++ b/arch/x86/kernel/cpu/bugs.c	2020-09-04 07:03:40.165031882 +0200
@@ -1525,12 +1525,12 @@
 	    (l1tf_vmx_mitigation == VMENTER_L1D_FLUSH_NEVER &&
 	     sched_smt_active())) {
 		return sprintf(buf, "%s; VMX: %s\n", L1TF_DEFAULT_MSG,
-			       l1tf_vmx_states[l1tf_vmx_mitigation]);
+			       l1tf_vmx_states[VMENTER_L1D_FLUSH_ALWAYS]);
 	}
 
 	return sprintf(buf, "%s; VMX: %s, SMT %s\n", L1TF_DEFAULT_MSG,
-		       l1tf_vmx_states[l1tf_vmx_mitigation],
-		       sched_smt_active() ? "vulnerable" : "disabled");
+		       l1tf_vmx_states[VMENTER_L1D_FLUSH_ALWAYS],
+		       "disabled");
 }
 
 static ssize_t itlb_multihit_show_state(char *buf)
