diff -uN a/drivers/input/touchscreen/surface3_spi.c b/drivers/input/touchscreen/surface3_spi.c
--- a/drivers/input/touchscreen/surface3_spi.c	2019-12-01 09:17:47.000000000 +0100
+++ b/drivers/input/touchscreen/surface3_spi.c	2020-01-22 20:28:16.653617439 +0100
@@ -21,6 +21,7 @@
 #include <linux/slab.h>
 #include <linux/spi/spi.h>
 #include <linux/acpi.h>
+#include <linux/surface_devices_dmi.h>
 
 #include <asm/unaligned.h>
 
@@ -29,6 +30,8 @@
 #define SURFACE3_REPORT_TOUCH	0xd2
 #define SURFACE3_REPORT_PEN	0x16
 
+static const struct dmi_system_id devices[] = surface_3_devices;
+
 struct surface3_ts_data {
 	struct spi_device *spi;
 	struct gpio_desc *gpiod_rst[2];
@@ -372,6 +375,9 @@
 	if (error)
 		return error;
 
+	if (dmi_check_system(devices))
+		spi->controller->can_dma = false;
+
 	return 0;
 }
 
