From 2903f0610336ac96411206c9458bd2885d11ea2d Mon Sep 17 00:00:00 2001
From: Alberto Milone <alberto.milone@canonical.com>
Date: Tue, 11 Dec 2012 15:51:26 +0100
Subject: [PATCH 1/2] Make sure that the module licence is "MIXED/Proprietary"

---
 src/wl/sys/wl_linux.c |    2 ++
 1 file changed, 2 insertions(+)

diff --git a/src/wl/sys/wl_linux.c b/src/wl/sys/wl_linux.c
index f64496b..af3bea6 100644
--- a/src/wl/sys/wl_linux.c
+++ b/src/wl/sys/wl_linux.c
@@ -159,6 +159,8 @@ static int wl_set_radio_block(void *data, bool blocked);
 static void wl_report_radio_state(wl_info_t *wl);
 #endif
 
+MODULE_LICENSE("MIXED/Proprietary");
+
 static struct pci_device_id wl_id_table[] =
 {
 	{ PCI_ANY_ID, PCI_ANY_ID, PCI_ANY_ID, PCI_ANY_ID,
-- 
1.7.9.5

