From 15bdf69c7c1fc0d674261b08a22c958e4826bc6b Mon Sep 17 00:00:00 2001
From: Alberto Milone <alberto.milone@canonical.com>
Date: Tue, 11 Dec 2012 15:55:08 +0100
Subject: [PATCH 2/2] Make sure to use the binary for the correct architecture

---
 Makefile |    8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index bf30b03..1f9b974 100644
--- a/Makefile
+++ b/Makefile
@@ -113,6 +113,12 @@ ifeq ($(APIFINAL),WEXT)
   $(info Using Wireless Extension API)
 endif
 
+ifeq ($(shell dpkg --print-architecture),amd64)
+BINARCH = x86_64
+else
+BINARCH = i386
+endif
+
 obj-m              += wl.o
 
 wl-objs            := 
@@ -125,7 +131,7 @@ EXTRA_CFLAGS       += -I$(src)/src/include
 EXTRA_CFLAGS       += -I$(src)/src/wl/sys -I$(src)/src/wl/clm/api -I$(src)/src/wl/phy
 #EXTRA_CFLAGS       += -DBCMDBG_ASSERT
 
-EXTRA_LDFLAGS      := $(src)/lib/wlc_hybrid.o_shipped
+EXTRA_LDFLAGS      := $(src)/lib/wlc_hybrid.o_shipped_$(BINARCH)
 
 KBASE              ?= /lib/modules/`uname -r`
 KBUILD_DIR         ?= $(KBASE)/build
-- 
1.7.9.5

