From 165f72491ffe0ffdf21c45ec7d5d29e16e2b6256 Mon Sep 17 00:00:00 2001
From: Alberto Milone <alberto.milone@canonical.com>
Date: Mon, 25 Mar 2013 14:08:51 +0100
Subject: [PATCH 1/1] Add support for Linux 3.9

---
--- a/src/wl/sys/wl_cfg80211_hybrid.c
+++ a/src/wl/sys/wl_cfg80211_hybrid.c
@@ -2048,7 +2048,11 @@
 		ie_len = (size_t)(ies->len);
 		rcu_read_unlock();
 #endif
+#if LINUX_VERSION_CODE < KERNEL_VERSION(3, 9, 0)
 		cfg80211_put_bss(bss);
+#else
+		cfg80211_put_bss(wl_to_wiphy(wl), bss);
+#endif
 	}
 
 	tim = bcm_parse_tlvs(ie, ie_len, WLAN_EID_TIM);
