--- a/drivers/firmware/dmi-id.c	2019-11-18 05:36:25.000000000 +0100
+++ b/drivers/firmware/dmi-id.c	2019-12-14 10:33:20.289505600 +0100
@@ -26,7 +26,12 @@
 {
 	int field = to_dmi_dev_attr(attr)->field;
 	ssize_t len;
-	len = scnprintf(page, PAGE_SIZE, "%s\n", dmi_get_system_info(field));
+	if (field == DMI_PRODUCT_NAME)
+		len = scnprintf(page, PAGE_SIZE, "Brunch\n");
+	else if (field == DMI_PRODUCT_SKU)
+		len = scnprintf(page, PAGE_SIZE, "sku0\n");
+	else
+		len = scnprintf(page, PAGE_SIZE, "%s\n", dmi_get_system_info(field));
 	page[len-1] = '\n';
 	return len;
 }
