diff -ruN a/drivers/net/wireless/marvell/mwifiex/sta_cmd.c b/drivers/net/wireless/marvell/mwifiex/sta_cmd.c
--- a/drivers/net/wireless/marvell/mwifiex/sta_cmd.c	2019-09-15 23:19:32.000000000 +0200
+++ b/drivers/net/wireless/marvell/mwifiex/sta_cmd.c	2019-10-29 07:37:26.664018907 +0100
@@ -17,6 +17,8 @@
  * this warranty disclaimer.
  */
 
+#include <linux/surface_devices_dmi.h>
+
 #include "decl.h"
 #include "ioctl.h"
 #include "util.h"
@@ -26,6 +28,8 @@
 #include "11n.h"
 #include "11ac.h"
 
+static const struct dmi_system_id devices[] = surface_mwifiex_pcie_devices;
+
 static bool drcs;
 module_param(drcs, bool, 0644);
 MODULE_PARM_DESC(drcs, "multi-channel operation:1, single-channel operation:0");
@@ -2395,7 +2399,7 @@
 	if (ret)
 		return -1;
 
-	if (!disable_auto_ds && first_sta &&
+	if (!dmi_check_system(devices) && !disable_auto_ds && first_sta &&
 	    priv->bss_type != MWIFIEX_BSS_TYPE_UAP) {
 		/* Enable auto deep sleep */
 		auto_ds.auto_ds = DEEP_SLEEP_ON;
