--- a/arch/x86/kernel/smpboot.c	2020-01-10 19:22:17.000000000 +0100
+++ b/arch/x86/kernel/smpboot.c	2020-03-13 18:41:38.757782269 +0100
@@ -1416,8 +1416,8 @@
 	if (setup_possible_cpus == -1) {
 		possible = num_processors;
 #ifdef CONFIG_HOTPLUG_CPU
-		if (setup_max_cpus)
-			possible += disabled_cpus;
+		//if (setup_max_cpus)
+		//	possible += disabled_cpus;
 #else
 		if (possible > i)
 			possible = i;
@@ -1425,7 +1425,8 @@
 	} else
 		possible = setup_possible_cpus;
 
-	total_cpus = max_t(int, possible, num_processors + disabled_cpus);
+	//total_cpus = max_t(int, possible, num_processors + disabled_cpus);
+	total_cpus = max_t(int, possible, num_processors);
 
 	/* nr_cpu_ids could be reduced via nr_cpus= */
 	if (possible > nr_cpu_ids) {
