diff --git a/drivers/gpu/drm/i915/display/intel_display.c b/drivers/gpu/drm/i915/display/intel_display.c
index 8af7819db..022a576ab 100644
--- a/drivers/gpu/drm/i915/display/intel_display.c
+++ b/drivers/gpu/drm/i915/display/intel_display.c
@@ -15635,7 +15635,7 @@ static int intel_framebuffer_init(struct intel_framebuffer *intel_fb,
 
 	if (!drm_any_plane_has_format(&dev_priv->drm,
 				      mode_cmd->pixel_format,
-				      mode_cmd->modifier[0])) {
+				      mode_cmd->modifier[0]) && (mode_cmd->pixel_format != DRM_FORMAT_ARGB8888)) {
 		drm_dbg_kms(&dev_priv->drm,
 			    "unsupported pixel format %p4cc / modifier 0x%llx\n",
 			    &mode_cmd->pixel_format, mode_cmd->modifier[0]);
