--- b/include/linux/etherdevice.h
+++ a/include/linux/etherdevice.h
@@ -299,18 +299,6 @@
 #endif
 }
 
-/**
- * eth_hw_addr_set - Assign Ethernet address to a net_device
- * @dev: pointer to net_device structure
- * @addr: address to assign
- *
- * Assign given address to the net_device, addr_assign_type is not changed.
- */
-static inline void eth_hw_addr_set(struct net_device *dev, const u8 *addr)
-{
-	ether_addr_copy(dev->dev_addr, addr);
-}
-
 /**
  * eth_hw_addr_inherit - Copy dev_addr from another net_device
  * @dst: pointer to net_device to copy dev_addr to
--- b/include/linux/netdevice.h
+++ a/include/linux/netdevice.h
@@ -4474,24 +4474,6 @@
 void __hw_addr_init(struct netdev_hw_addr_list *list);
 
 /* Functions used for device addresses handling */
-static inline void
-__dev_addr_set(struct net_device *dev, const u8 *addr, size_t len)
-{
-	memcpy(dev->dev_addr, addr, len);
-}
-
-static inline void dev_addr_set(struct net_device *dev, const u8 *addr)
-{
-	__dev_addr_set(dev, addr, dev->addr_len);
-}
-
-static inline void
-dev_addr_mod(struct net_device *dev, unsigned int offset,
-	     const u8 *addr, size_t len)
-{
-	memcpy(&dev->dev_addr[offset], addr, len);
-}
-
 int dev_addr_add(struct net_device *dev, const unsigned char *addr,
 		 unsigned char addr_type);
 int dev_addr_del(struct net_device *dev, const unsigned char *addr,
