--- a/drivers/firmware/dmi_scan.c	2020-01-10 19:22:21.000000000 +0100
+++ b/drivers/firmware/dmi_scan.c	2020-01-13 21:06:59.580077401 +0100
@@ -614,6 +614,17 @@
 	return 1;
 }
 
+static void __init dmi_surface3_workaround(void)
+{
+	const char *surface3_product_name = "Surface 3";
+	const char *surface3_sys_vendor = "Microsoft Corporation";
+
+	if (!strncmp(dmi_ident[DMI_PRODUCT_NAME], "OEMB", 4) && !strncmp(dmi_ident[DMI_SYS_VENDOR], "OEMB", 4)) {
+		dmi_ident[DMI_PRODUCT_NAME] = surface3_product_name;
+		dmi_ident[DMI_SYS_VENDOR] = surface3_sys_vendor;
+	}
+}
+
 void __init dmi_scan_machine(void)
 {
 	char __iomem *p, *q;
@@ -642,6 +653,7 @@
 
 		if (!dmi_present(buf)) {
 			dmi_available = 1;
+			dmi_surface3_workaround();
 			return;
 		}
 	} else if (IS_ENABLED(CONFIG_DMI_SCAN_MACHINE_NON_EFI_FALLBACK)) {
