diff -ruN a/drivers/pci/pci-acpi.c b/drivers/pci/pci-acpi.c
--- a/drivers/pci/pci-acpi.c	2019-09-15 23:19:32.000000000 +0200
+++ b/drivers/pci/pci-acpi.c	2019-10-31 19:06:37.703337211 +0100
@@ -18,8 +18,11 @@
 #include <linux/pci-acpi.h>
 #include <linux/pm_runtime.h>
 #include <linux/pm_qos.h>
+#include <linux/surface_devices_dmi.h>
 #include "pci.h"
 
+static const struct dmi_system_id devices[] = surface_all_devices;
+
 /*
  * The GUID is defined in the PCI Firmware Specification available here:
  * https://www.pcisig.com/members/downloads/pcifw_r3_1_13Dec10.pdf
@@ -912,6 +915,11 @@
 
 	acpi_pci_wakeup(pci_dev, false);
 	acpi_device_power_add_dependent(adev, dev);
+
+	if (dmi_check_system(devices) &&
+	    pci_dev->bus->self && pci_is_pcie(pci_dev->bus->self) &&
+	    pci_pcie_type(pci_dev->bus->self) == PCI_EXP_TYPE_ROOT_PORT)
+		adev->power.flags.ignore_parent = true;
 }
 
 static void pci_acpi_cleanup(struct device *dev)
