diff -uN a/include/linux/surface_devices_dmi.h b/include/linux/surface_devices_dmi.h
--- a/include/linux/surface_devices_dmi.h	1970-01-01 01:00:00.000000000 +0100
+++ b/include/linux/surface_devices_dmi.h	2020-01-28 13:25:34.639217068 +0100
@@ -0,0 +1,122 @@
+/* SPDX-License-Identifier: GPL-2.0 */
+/*
+ * Macros to simplify Microsoft Surface devices specific patches.
+ */
+
+#ifndef _LINUX_SURFACE_DEVICES_DMI_H_
+#define _LINUX_SURFACE_DEVICES_DMI_H_
+
+#include <linux/dmi.h>
+
+#define surface_all_devices \
+	{ \
+		{ \
+			.ident = "Microsoft Surface devices", \
+			.matches = { \
+				DMI_EXACT_MATCH(DMI_SYS_VENDOR, "Microsoft Corporation"), \
+				DMI_MATCH(DMI_PRODUCT_NAME, "Surface"), \
+			}, \
+		}, \
+		{} \
+	}
+
+#define surface_3_devices \
+	{ \
+		{ \
+			.ident = "Microsoft Surface 3", \
+			.matches = { \
+				DMI_EXACT_MATCH(DMI_SYS_VENDOR, "Microsoft Corporation"), \
+				DMI_EXACT_MATCH(DMI_PRODUCT_NAME, "Surface 3"), \
+			}, \
+		}, \
+		{} \
+	}
+
+#define surface_mwifiex_pcie_devices \
+	{ \
+		{ \
+			.ident = "Microsoft Surface 3", \
+			.matches = { \
+				DMI_EXACT_MATCH(DMI_SYS_VENDOR, "Microsoft Corporation"), \
+				DMI_EXACT_MATCH(DMI_PRODUCT_NAME, "Surface 3"), \
+			}, \
+		}, \
+		{ \
+			.ident = "Microsoft Surface Pro 3", \
+			.matches = { \
+				DMI_EXACT_MATCH(DMI_SYS_VENDOR, "Microsoft Corporation"), \
+				DMI_EXACT_MATCH(DMI_PRODUCT_NAME, "Surface Pro 3"), \
+			}, \
+		}, \
+		{ \
+			.ident = "Microsoft Surface Pro 4", \
+			.matches = { \
+				DMI_EXACT_MATCH(DMI_SYS_VENDOR, "Microsoft Corporation"), \
+				DMI_EXACT_MATCH(DMI_PRODUCT_NAME, "Surface Pro 4"), \
+			}, \
+		}, \
+		{ \
+			.ident = "Microsoft Surface Pro 2017", \
+			.matches = { \
+				DMI_EXACT_MATCH(DMI_SYS_VENDOR, "Microsoft Corporation"), \
+				DMI_EXACT_MATCH(DMI_PRODUCT_NAME, "Surface Pro"), \
+			}, \
+		}, \
+		{ \
+			.ident = "Microsoft Surface Pro 6", \
+			.matches = { \
+				DMI_EXACT_MATCH(DMI_SYS_VENDOR, "Microsoft Corporation"), \
+				DMI_EXACT_MATCH(DMI_PRODUCT_NAME, "Surface Pro 6"), \
+			}, \
+		}, \
+		{ \
+			.ident = "Microsoft Surface Book 1", \
+			.matches = { \
+				DMI_EXACT_MATCH(DMI_SYS_VENDOR, "Microsoft Corporation"), \
+				DMI_EXACT_MATCH(DMI_PRODUCT_NAME, "Surface Book"), \
+			}, \
+		}, \
+		{ \
+			.ident = "Microsoft Surface Book 2", \
+			.matches = { \
+				DMI_EXACT_MATCH(DMI_SYS_VENDOR, "Microsoft Corporation"), \
+				DMI_EXACT_MATCH(DMI_PRODUCT_NAME, "Surface Book 2"), \
+			}, \
+		}, \
+		{ \
+			.ident = "Microsoft Surface Laptop 1", \
+			.matches = { \
+				DMI_EXACT_MATCH(DMI_SYS_VENDOR, "Microsoft Corporation"), \
+				DMI_EXACT_MATCH(DMI_PRODUCT_NAME, "Surface Laptop"), \
+			}, \
+		}, \
+		{ \
+			.ident = "Microsoft Surface Laptop 2", \
+			.matches = { \
+				DMI_EXACT_MATCH(DMI_SYS_VENDOR, "Microsoft Corporation"), \
+				DMI_EXACT_MATCH(DMI_PRODUCT_NAME, "Surface Laptop 2"), \
+			}, \
+		}, \
+		{ \
+			.ident = "Microsoft Surface Studio 1", \
+			.matches = { \
+				DMI_EXACT_MATCH(DMI_SYS_VENDOR, "Microsoft Corporation"), \
+				DMI_EXACT_MATCH(DMI_PRODUCT_NAME, "Surface Studio"), \
+			}, \
+		}, \
+		{} \
+	}
+
+#define surface_go_devices \
+	{ \
+		{ \
+			.ident = "Microsoft Surface Go 1", \
+			.matches = { \
+				DMI_EXACT_MATCH(DMI_SYS_VENDOR, "Microsoft Corporation"), \
+				DMI_EXACT_MATCH(DMI_PRODUCT_NAME, "Surface Go"), \
+			}, \
+		}, \
+		{} \
+	}
+
+#endif /* _LINUX_SURFACE_DEVICES_DMI_H_ */
