--- a/drivers/gpu/drm/drm_panel_orientation_quirks.c	2020-05-28 14:20:28.000000000 +0200
+++ b/drivers/gpu/drm/drm_panel_orientation_quirks.c	2020-06-27 16:03:26.865897348 +0200
@@ -94,6 +94,24 @@
 	.orientation = DRM_MODE_PANEL_ORIENTATION_RIGHT_UP,
 };
 
+static const struct drm_dmi_panel_orientation_data lcd800x1280_bottom_up = {
+	.width = 800,
+	.height = 1280,
+	.orientation = DRM_MODE_PANEL_ORIENTATION_BOTTOM_UP,
+};
+
+static const struct drm_dmi_panel_orientation_data lcd1200x1920_rightside_up = {
+	.width = 1200,
+	.height = 1920,
+	.orientation = DRM_MODE_PANEL_ORIENTATION_RIGHT_UP,
+};
+
+static const struct drm_dmi_panel_orientation_data lcd1280x1920_leftside_up = {
+	.width = 1280,
+	.height = 1920,
+	.orientation = DRM_MODE_PANEL_ORIENTATION_LEFT_UP,
+};
+
 static const struct dmi_system_id orientation_data[] = {
 	{	/* Acer One 10 (S1003) */
 		.matches = {
@@ -158,6 +164,37 @@
 		  DMI_EXACT_MATCH(DMI_BOARD_NAME, "TW891"),
 		},
 		.driver_data = (void *)&itworks_tw891,
+	}, {	/* EVOO EV-T2in1-101-2 */
+		.matches = {
+		  DMI_EXACT_MATCH(DMI_SYS_VENDOR, "EVOO Products Company, LLC."),
+		  DMI_EXACT_MATCH(DMI_PRODUCT_NAME, "EV-T2in1-101-2"),
+		},
+		.driver_data = (void *)&lcd800x1280_bottom_up,
+	}, {	/* CHUWI Innovation And Technology(ShenZhen)co.,Ltd Hi10 X */
+		.matches = {
+		  DMI_EXACT_MATCH(DMI_SYS_VENDOR, "CHUWI Innovation And Technology(ShenZhen)co.,Ltd"),
+		  DMI_EXACT_MATCH(DMI_PRODUCT_NAME, "Hi10 X"),
+		},
+		.driver_data = (void *)&lcd1200x1920_rightside_up,
+	}, {	/* CHUWI INNOVATION AND TECHNOLOGY(SHENZHEN)CO.LTD Hi10 pro tablet */
+		.matches = {
+		  DMI_EXACT_MATCH(DMI_SYS_VENDOR, "CHUWI INNOVATION AND TECHNOLOGY(SHENZHEN)CO.LTD"),
+		  DMI_EXACT_MATCH(DMI_PRODUCT_NAME, "Hi10 pro tablet"),
+		},
+		.driver_data = (void *)&lcd1200x1920_rightside_up,
+	}, {	/* Lenovo Ideapad D330 */
+		.matches = {
+		  DMI_EXACT_MATCH(DMI_SYS_VENDOR, "LENOVO"),
+		  DMI_EXACT_MATCH(DMI_PRODUCT_NAME, "81H3"),
+		  DMI_EXACT_MATCH(DMI_PRODUCT_VERSION, "Lenovo ideapad D330-10IGM"),
+		},
+		.driver_data = (void *)&lcd800x1280_rightside_up,
+	}, {	/* Samsung Galaxy Book 10.6 */
+		.matches = {
+		  DMI_EXACT_MATCH(DMI_SYS_VENDOR, "SAMSUNG ELECTRONICS CO., LTD."),
+		  DMI_EXACT_MATCH(DMI_PRODUCT_NAME, "Galaxy Book 10.6"),
+		},
+		.driver_data = (void *)&lcd1280x1920_leftside_up,
 	}, {	/* KD Kurio Smart C15200 2-in-1 */
 		.matches = {
 		  DMI_EXACT_MATCH(DMI_SYS_VENDOR, "KD Interactive"),
@@ -211,26 +230,34 @@
 	const struct drm_dmi_panel_orientation_data *data;
 	const char *bios_date;
 	int i;
+	
+	pr_info("drm_get_panel_orientation_quirk called with width=%d height=%d", width, height);
 
 	for (match = dmi_first_match(orientation_data);
 	     match;
 	     match = dmi_first_match(match + 1)) {
 		data = match->driver_data;
+		
+		pr_info("drm_get_panel_orientation_quirk dmi match found");
 
 		if (data->width != width ||
 		    data->height != height)
 			continue;
 
-		if (!data->bios_dates)
+		if (!data->bios_dates) {
+			pr_info("drm_get_panel_orientation_quirk dmi match applied");
 			return data->orientation;
+		}
 
 		bios_date = dmi_get_system_info(DMI_BIOS_DATE);
 		if (!bios_date)
 			continue;
 
 		i = match_string(data->bios_dates, -1, bios_date);
-		if (i >= 0)
+		if (i >= 0) {
+			pr_info("drm_get_panel_orientation_quirk dmi match applied");
 			return data->orientation;
+		}
 	}
 
 	return DRM_MODE_PANEL_ORIENTATION_UNKNOWN;
