diff --git a/kernel/drivers/hid/hid-input.c b/kernel/drivers/hid/hid-input.c
index ba0a8fc58..a888aaea0 100644
--- a/drivers/hid/hid-input.c
+++ b/drivers/hid/hid-input.c
@@ -1307,11 +1307,14 @@
 
 	if (usage->hid == (HID_UP_DIGITIZER | 0x0032)) { /* InRange */
 		if (value) {
+                       if ((*quirks & HID_QUIRK_INVERT))
+                           input_event(input, usage->type, BTN_TOOL_PEN, 0);
 			input_event(input, usage->type, (*quirks & HID_QUIRK_INVERT) ? BTN_TOOL_RUBBER : usage->code, 1);
 			return;
 		}
 		input_event(input, usage->type, usage->code, 0);
 		input_event(input, usage->type, BTN_TOOL_RUBBER, 0);
+               input_event(input, usage->type, BTN_TOOL_PEN, 0);
 		return;
 	}
 
