--- a/drivers/acpi/acpica/dswexec.c	2022-03-24 15:15:58.470866679 +0100
+++ b/drivers/acpi/acpica/dswexec.c	2022-03-24 15:14:27.497035303 +0100
@@ -391,7 +391,8 @@
 		 * All opcodes require operand resolution, with the only exceptions
 		 * being the object_type and size_of operators.
 		 */
-		if (!(walk_state->op_info->flags & AML_NO_OPERAND_RESOLVE)) {
+		if (!(walk_state->op_info->flags & AML_NO_OPERAND_RESOLVE) &&
+			walk_state->num_operands > 0) {

 			/* Resolve all operands */
