--- a/security/selinux/include/security.h	2021-02-17 11:02:30.000000000 +0100
+++ b/security/selinux/include/security.h	2021-03-27 23:56:33.517341787 +0100
@@ -125,12 +126,11 @@
 #ifdef CONFIG_SECURITY_SELINUX_DEVELOP
 static inline bool enforcing_enabled(struct selinux_state *state)
 {
-	return READ_ONCE(state->enforcing);
+	return false;
 }
 
 static inline void enforcing_set(struct selinux_state *state, bool value)
 {
-	WRITE_ONCE(state->enforcing, value);
 }
 #else
 static inline bool enforcing_enabled(struct selinux_state *state)
--- a/security/selinux/selinuxfs.c	2020-05-24 11:17:09.000000000 +0200
+++ b/security/selinux/selinuxfs.c	2020-06-27 08:23:35.906531847 +0200
@@ -125,8 +125,9 @@
 	char tmpbuf[TMPBUFLEN];
 	ssize_t length;
 
-	length = scnprintf(tmpbuf, TMPBUFLEN, "%d",
-			   enforcing_enabled(fsi->state));
+	//length = scnprintf(tmpbuf, TMPBUFLEN, "%d",
+	//		   enforcing_enabled(fsi->state));
+	length = scnprintf(tmpbuf, TMPBUFLEN, "1");
 	return simple_read_from_buffer(buf, count, ppos, tmpbuf, length);
 }
 
