--- a/drivers/base/core.c	2020-05-28 14:20:19.000000000 +0200
+++ b/drivers/base/core.c	2020-05-30 09:02:19.896003755 +0200
@@ -43,6 +43,13 @@
 early_param("sysfs.deprecated", sysfs_deprecated_setup);
 #endif
 
+long enforce_hyperthreading = 0;
+static int __init enforce_hyperthreading_setup(char *arg)
+{
+	return kstrtol(arg, 10, &enforce_hyperthreading);
+}
+early_param("enforce_hyperthreading", enforce_hyperthreading_setup);
+
 /* Device links support. */
 
 #ifdef CONFIG_SRCU
@@ -1267,6 +1274,9 @@
 	bool val;
 	int ret;
 
+	if (enforce_hyperthreading)
+		return count;
+
 	ret = strtobool(buf, &val);
 	if (ret < 0)
 		return ret;
