--- a/drivers/platform/x86/intel-hid.c	2020-04-20 19:28:32.000000000 +0200
+++ b/drivers/platform/x86/intel-hid.c	2020-05-04 18:01:26.427423389 +0200
@@ -60,6 +60,20 @@
 
 static const struct dmi_system_id button_array_table[] = {
 	{
+		.ident = "HUAWEI MateBook HZ-W09",
+		.matches = {
+			DMI_MATCH(DMI_SYS_VENDOR, "HUAWEI"),
+			DMI_MATCH(DMI_PRODUCT_NAME, "MateBook HZ-W09"),
+		},
+	},
+	{
+		.ident = "HUAWEI MateBook HZ-W19",
+		.matches = {
+			DMI_MATCH(DMI_SYS_VENDOR, "HUAWEI"),
+			DMI_MATCH(DMI_PRODUCT_NAME, "MateBook HZ-W19"),
+		},
+	},
+	{
 		.ident = "Wacom MobileStudio Pro 13",
 		.matches = {
 			DMI_MATCH(DMI_SYS_VENDOR, "Wacom Co.,Ltd"),
