--- a/drivers/gpu/drm/drm_panel_orientation_quirks.c	2020-04-20 19:28:13.000000000 +0200
+++ b/drivers/gpu/drm/drm_panel_orientation_quirks.c	2020-05-21 19:00:15.891949179 +0200
@@ -15,6 +15,17 @@
 
 #ifdef CONFIG_DMI
 
+static const struct dmi_system_id inverted_screen[] = {
+	{	/* TrekStor SurfTab duo W3 */
+        	.ident = "TrekStor SurfTab duo W3",
+		.matches = {
+		  DMI_MATCH(DMI_SYS_VENDOR, "TrekStor"),
+		  DMI_MATCH(DMI_PRODUCT_NAME, "SurfTab duo W3"),
+		},
+	},
+	{}
+};
+
 /*
  * Some x86 clamshell design devices use portrait tablet screens and a display
  * engine which cannot rotate in hardware, so we need to rotate the fbcon to
@@ -212,6 +223,11 @@
 	const char *bios_date;
 	int i;
 
+	if (dmi_check_system(inverted_screen)) {
+		pr_info("applying orientation quirk\n");
+		return DRM_MODE_PANEL_ORIENTATION_BOTTOM_UP;
+	}
+
 	for (match = dmi_first_match(orientation_data);
 	     match;
 	     match = dmi_first_match(match + 1)) {
