--- a/arch/x86/kernel/cpu/common.c	2021-05-22 16:34:50.000000000 +0200
+++ b/arch/x86/kernel/cpu/common.c	2021-06-19 08:39:15.613499390 +0200
@@ -1634,12 +1634,13 @@
 #endif /* ! CONFIG_KGDB */
 
 #ifdef CONFIG_CHROMEOS
-static int disablevmx = 1;
+static int disablevmx = 0;
 static int __init dodisablevmx(char *value)
 {
 	if (!value)
 		return 0;
 	if (!strncmp(value, "on", 2))
+		disablevmx = 1;
 		return 0;
 	if (!strncmp(value, "off", 3)) {
 		disablevmx = 0;
