--- a/drivers/firmware/efi/efi.c	2020-08-29 06:32:44.000000000 +0200
+++ b/drivers/firmware/efi/efi.c	2020-09-04 07:03:40.165031882 +0200
@@ -405,7 +405,7 @@
 		platform_device_register_simple("rtc-efi", 0, NULL, 0);
 
 	/* We register the efi directory at /sys/firmware/efi */
-	efi_kobj = kobject_create_and_add("efi", firmware_kobj);
+	efi_kobj = kobject_create_and_add("brunchefi", firmware_kobj);
 	if (!efi_kobj) {
 		pr_err("efi: Firmware registration failed.\n");
 		error = -ENOMEM;
