--- a/drivers/gpu/drm/i915/i915_params.h	2020-04-20 19:28:13.000000000 +0200
+++ b/drivers/gpu/drm/i915/i915_params.h	2020-05-01 10:39:41.889124645 +0200
@@ -42,9 +42,9 @@
 	param(int, panel_use_ssc, -1) \
 	param(int, vbt_sdvo_panel_type, -1) \
 	param(int, enable_dc, -1) \
-	param(int, enable_fbc, -1) \
+	param(int, enable_fbc, 0) \
 	param(int, enable_ppgtt, -1) \
-	param(int, enable_psr, -1) \
+	param(int, enable_psr, 0) \
 	param(int, disable_power_well, -1) \
 	param(int, enable_ips, 1) \
 	param(int, invert_brightness, 0) \
--- a/drivers/tty/vt/vt.c	2020-04-20 19:28:38.000000000 +0200
+++ b/drivers/tty/vt/vt.c	2020-05-01 08:03:36.059292624 +0200
@@ -3411,7 +3411,7 @@
 		panic("Couldn't register /dev/tty0 driver\n");
 	tty0dev = device_create_with_groups(tty_class, NULL,
 					    MKDEV(TTY_MAJOR, 0), NULL,
-					    vt_dev_groups, "tty0");
+					    vt_dev_groups, "brunch-tty0");
 	if (IS_ERR(tty0dev))
 		tty0dev = NULL;
 
