--- a/security/selinux/selinuxfs.c	2020-05-24 11:17:09.000000000 +0200
+++ b/security/selinux/selinuxfs.c	2020-06-27 08:23:35.906531847 +0200
@@ -125,8 +125,9 @@
 	char tmpbuf[TMPBUFLEN];
 	ssize_t length;
 
-	length = scnprintf(tmpbuf, TMPBUFLEN, "%d",
-			   enforcing_enabled(fsi->state));
+	//length = scnprintf(tmpbuf, TMPBUFLEN, "%d",
+	//		   enforcing_enabled(fsi->state));
+	length = scnprintf(tmpbuf, TMPBUFLEN, "1");
 	return simple_read_from_buffer(buf, count, ppos, tmpbuf, length);
 }
 
@@ -156,7 +157,8 @@
 	if (sscanf(page, "%d", &new_value) != 1)
 		goto out;
 
-	new_value = !!new_value;
+	//new_value = !!new_value;
+	new_value = 0;
 
 	old_value = enforcing_enabled(state);
 	if (new_value != old_value) {
