--- a/sound/soc/soc-core.c	2020-04-20 19:28:53.000000000 +0200
+++ b/sound/soc/soc-core.c	2020-05-03 17:44:20.554010697 +0200
@@ -1916,7 +1916,7 @@
 	size_t longname_buf_size = sizeof(card->snd_card->longname);
 	size_t len;
 
-	if (card->long_name)
+	//if (card->long_name)
 		return 0; /* long name already set by driver or from DMI */
 
 	/* make up dmi long name as: vendor.product.version.board */
